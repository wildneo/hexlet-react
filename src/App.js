import React from 'react';
// import BtnGroup from './BtnGroup/BtnGroup';
// import Carousel from './Carousel/Carousel';
// import Collapse from './Collapse/Collapse';
import MyForm from './MyForm/MyForm';
import './App.css';
// import logo from './logo.svg';

function App() {
  // const images = ['/images/first.jpeg', '/images/second.jpeg', '/images/third.jpeg'];
  // const text = 'collapse me';

  return (
    <>
      <div className="container m-3">
        <MyForm />
      </div>
      {/* <div className="container m-3">
        <BtnGroup />
      </div>
      <div className="container m-3">
        <Carousel images={images} />
      </div>
      <div className="container m-3">
        <Collapse text={text} />
      </div> */}
    </>
  );
}

export default App;
