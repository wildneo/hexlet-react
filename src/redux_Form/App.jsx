import React from 'react';

export default class App extends React.Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleClick(event) {
    event.preventDefault();
    const { dispatch, reset } = this.props;
    dispatch(reset());
  }

  handleChange(event) {
    event.preventDefault();
    const { dispatch, change } = this.props;
    const { target: { value } } = event;
    dispatch(change(value));
  }

  render() {
    const { text } = this.props;

    return (
      <div>
        <form>
          <input onChange={this.handleChange} type="text" value={text} />
          <button onClick={this.handleClick} type="button">Reset</button>
        </form>
        {text && <div>{text}</div>}
      </div>
    );
  }
}

App.defaultProps = {
  text: '',
};
