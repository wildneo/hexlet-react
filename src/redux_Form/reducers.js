import { combineReducers } from 'redux';

const text = (state = '', action) => {
  switch (action.type) {
    case 'TEXT_CHANGE':
      return action.payload.value;

    case 'TEXT_RESET': {
      return '';
    }

    default:
      return state;
  }
};

export default combineReducers({
  text,
});
