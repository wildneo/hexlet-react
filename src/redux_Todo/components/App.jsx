import React from 'react';
import NewTaskForm from './NewTaskForm';
import Panel from './Panel';
import Tasks from './Tasks';

export default () => (
  <div className="col-5">
    <NewTaskForm />
    <Panel />
    <Tasks />
  </div>
);
