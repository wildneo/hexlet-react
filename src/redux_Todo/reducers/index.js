import { combineReducers } from 'redux';

const text = (state = '', action) => {
  switch (action.type) {
    case 'TEXT_UPDATE':
      return action.payload.text;

    case 'TASK_ADD':
      return '';

    default:
      return state;
  }
};

const tasks = (state = [], action) => {
  switch (action.type) {
    case 'TASK_ADD': {
      const { task } = action.payload;
      return [task, ...state];
    }

    case 'TASK_REMOVE': {
      const { id } = action.payload;
      return state.filter((task) => task.id !== id);
    }

    case 'TASK_CLEAN':
      return [];

    default:
      return state;
  }
};

export default combineReducers({
  text,
  tasks,
});
