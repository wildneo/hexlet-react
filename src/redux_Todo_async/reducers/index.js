import _ from 'lodash';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import { reducer as formReducer } from 'redux-form';
import * as actions from '../actions';

// BEGIN (write your solution here)
const taskRemovingState = handleActions({
  [actions.fetchTasksRequest]: () => 'requested',
  [actions.fetchTasksFailure]: () => 'failed',
  [actions.fetchTasksSuccess]: () => 'finished',
}, 'none');
// END

const tasksFetchingState = handleActions({
  [actions.fetchTasksRequest]: () => 'requested',
  [actions.fetchTasksFailure]: () => 'failed',
  [actions.fetchTasksSuccess]: () => 'finished',
}, 'none');

const tasks = handleActions({
  [actions.fetchTasksSuccess]: (state, { payload }) => ({
    byId: _.keyBy(payload.tasks, 'id'),
    allIds: payload.tasks.map((t) => t.id),
  }),

  [actions.addTaskSuccess]: (state, { payload: { task } }) => {
    const { byId, allIds } = state;

    return {
      byId: { ...byId, [task.id]: task },
      allIds: [task.id, ...allIds],
    };
  },

  [actions.removeTaskSuccess]: (state, { payload: { id } }) => {
    const { byId, allIds } = state;

    return {
      byId: _.omit(byId, id),
      allIds: _.without(allIds, id),
    };
  },
}, { byId: {}, allIds: [] });

export default combineReducers({
  taskRemovingState,
  tasksFetchingState,
  tasks,
  form: formReducer,
});
