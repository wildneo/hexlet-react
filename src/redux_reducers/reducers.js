import _ from 'lodash';
import { combineReducers } from 'redux';

const tasks = (state = {}, action) => {
  switch (action.type) {
    case 'TASK_ADD': {
      const { task } = action.payload;

      return { ...state, [task.id]: task };
    }

    case 'TASK_REMOVE': {
      const { id } = action.payload;

      return _.omit(state, id);
    }

    default:
      return state;
  }
};

const comments = (state = {}, action) => {
  switch (action.type) {
    case 'TASK_REMOVE': {
      const { id } = action.payload;

      return _.omitBy(state, (c) => c.taskId === id);
    }

    case 'TASK_COMMENT_ADD': {
      const { comment } = action.payload;

      return { ...state, [comment.id]: comment };
    }

    case 'TASK_COMMENT_REMOVE': {
      const { id } = action.payload;

      return _.omit(state, id);
    }

    default:
      return state;
  }
};

export default combineReducers({ tasks, comments });
