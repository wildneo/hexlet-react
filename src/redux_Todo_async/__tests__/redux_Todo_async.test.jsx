import timeout from 'timeout-then';
import React from 'react';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import axios from 'axios';
import thunk from 'redux-thunk';
import httpAdapter from 'axios/lib/adapters/http';
import nock from 'nock';
import App from '../components/App';
import reducers from '../reducers';

Enzyme.configure({ adapter: new Adapter() });

axios.defaults.adapter = httpAdapter;
nock.disableNetConnect();

const host = 'http://localhost';

test('Store', async () => {
  const store = createStore(
    reducers,
    applyMiddleware(thunk),
  );

  const vdom = (
    <Provider store={store}>
      <App />
    </Provider>
  );
  const wrapper = mount(vdom);
  expect(wrapper.render()).toMatchSnapshot();

  const newTaskInput = wrapper.find('input[type="text"]');
  const newTaskSubmit = wrapper.find('input[type="submit"]');

  newTaskInput.simulate('change', { target: { value: 'na-na' } });
  expect(wrapper.render()).toMatchSnapshot();

  nock(host)
    .post('/tasks')
    .reply(201, (uri, body) => {
      const response = { ...body.task, state: 'active', id: 1 };
      return response;
    });

  newTaskSubmit.simulate('submit');
  await timeout(100);
  expect(wrapper.render()).toMatchSnapshot();

  newTaskInput.simulate('change', { target: { value: 'another task' } });
  expect(wrapper.render()).toMatchSnapshot();

  nock(host).post('/tasks')
    .reply(201, (uri, body) => {
      const response = { ...body.task, state: 'active', id: 2 };
      return response;
    });

  newTaskSubmit.simulate('submit');
  await timeout(100);
  expect(wrapper.render()).toMatchSnapshot();

  const links = wrapper.find('[data-test="task-remove"]');
  nock(host).delete('/tasks/1')
    .reply(204);
  links.last().simulate('click');
  await timeout(100);
  expect(wrapper.render()).toMatchSnapshot();
});
