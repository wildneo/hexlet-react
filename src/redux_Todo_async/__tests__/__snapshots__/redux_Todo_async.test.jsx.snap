// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store 1`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
      value="Add"
    />
  </form>
</div>
`;

exports[`Store 2`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value="na-na"
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
</div>
`;

exports[`Store 3`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          data-test="task-remove"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 4`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value="another task"
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          data-test="task-remove"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 5`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          another task
        </span>
        <button
          class="close"
          data-test="task-remove"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          data-test="task-remove"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 6`] = `
<div
  class="col-5"
>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          another task
        </span>
        <button
          class="close"
          data-test="task-remove"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
